go-controller/pkg/level_driven_controller/level_driven_controller.go

```diff
@@ -0,0 +1,271 @@
+	}
+
+	for i := 0; i < threadiness; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			// this loop will exit once the queue is shut down, because processNextQueueItem will return false
+			for c.processNextQueueItem() {
+			}
+		}()
+	}
+
+	// add shutdown goroutine waiting for c.stopCh
+	go func() {
+		// wait until we're told to stop
+		<-stopCh
+		c.cleanup()
+	}()
+
+	return nil
+}
+
+func (c *controller[T]) cleanup() {
+	c.queue.ShutDown()
+	if c.eventHandler != nil {
+		if err := c.informer.RemoveEventHandler(c.eventHandler); err != nil {
+			klog.Errorf("Failed to remove event handler for controller %s: %w", c.name, err)
+		}
+	}
+}
+
+func (c *controller[T]) onAdd(objInterface interface{}) {
+	newObj, ok := objInterface.(*T)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("controller %s: expecting %T but received %T", c.name, *new(T), newObj))
+		return
+	}
+	if !c.handler.ObjNeedsUpdate(nil, newObj) {
+		return
+	}
+
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("controller %s: couldn't get key for object %+v: %v", c.name, newObj, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *controller[T]) onUpdate(oldObjInterface, newObjInterface interface{}) {
+	oldObj, ok := oldObjInterface.(*T)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("controller %s: expecting %T but received %T", c.name, *new(T), oldObj))
+		return
+	}
+	newObj, ok := newObjInterface.(*T)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("controller %s: expecting %T but received %T", c.name, *new(T), newObj))
+		return
+	}
+
+	if !c.handler.ObjNeedsUpdate(oldObj, newObj) {
+		return
+	}
+
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("controller %s: couldn't get key for object %+v: %v", c.name, newObj, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *controller[T]) onDelete(objInterface interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(objInterface)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("controller %s: couldn't get key for object %+v: %v", c.name, objInterface, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+// processNextQueueItem returns false when the queue is shutdown and the handling should be stopped.
+// Otherwise, it handles the next item from the queue and always returns true.
+func (c *controller[T]) processNextQueueItem() bool {
+	key, shutdown := c.queue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	defer c.queue.Done(key)
+
+	err := c.handler.Reconcile(key.(string))
+	if err != nil {
+		if c.queue.NumRequeues(key) < maxRetries {
+			klog.Infof("Controller %s: error found while processing %s: %v", c.name, key.(string), err)
+			c.queue.AddRateLimited(key)
+			return true
+		}
+		klog.Warningf("Controller %s: dropping %s out of the queue: %v", c.name, key.(string), err)
+		utilruntime.HandleError(err)
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *controller[T]) AddToQueue(key string) {
```

Comment:
It is used by status manager when zones change to reconcile all objects. I think generally speaking something like this makes sense, but I am not sure what is better `AddToQueue` or `ReconcileAll`

Reply:
OK. 

So this is a way to kind of connect or compose a sub-controllers into a single controller. Just as the library I pointed before, same capability but a slightly different approach, where you have a single controller that can watch multiple things.

Might look a bit weird to people accustomed to see traditional controllers, and who knows what kind of complications this might lead us, but nothing we haven't done before so I guess this is how we will roll for now.

It looks like ReconcileAll could be done here and It would save of us from some code repetition?